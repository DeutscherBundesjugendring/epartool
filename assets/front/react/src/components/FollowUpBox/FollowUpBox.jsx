import React from 'react';
import PropTypes from 'prop-types';
import ArrowButton from '../ArrowButton/ArrowButton';
import RaisedButton from '../RaisedButton/RaisedButton';

class FollowUpBox extends React.Component {
  componentDidMount() {
    this.adjustContributionHeight();
    this.addToggle();
  }

  componentDidUpdate() {
    this.adjustContributionHeight();
    this.addToggle();
  }

  adjustContributionHeight() {
    /*
    Makes second div in contributionBox shrink to accommodate first div
    Lives in this component instead of contributionBox b/c of forceUpdate
    Operating on DOM b/c needs to happen after render
    */
    const {
      id,
      isOpened,
      type,
    } = this.props;

    const wellArray = window.document
      .getElementById(`${type}-${id}`)
      .getElementsByClassName('followup-flow')[0];
    if (type === 'contribution') {
      // 400px is collapsed followup-timeline-box height, -20px is adjustment for margin
      const newHeight = 400
        - (wellArray.getElementsByClassName('js-followup-box-head')[0].offsetHeight - 20);
      wellArray.getElementsByClassName('js-followup-box-content')[0].style.height = isOpened() ? '' : `${newHeight}px`;
    } else {
      const headEl = wellArray.getElementsByClassName('js-followup-box-head')[0];
      // 400px is collapsed followup-timeline-box height, -20px is adjustment for margin
      // 31.01.2018 - jiri@visionapps.cz - 121px is an experimentally found value to prevent two
      // phase resizing of the box when it is collapsing
      const newHeight = 400 - headEl.offsetHeight - 121;
      wellArray.getElementsByClassName('js-followup-box-content')[0].style.height = isOpened()
        ? ''
        : `${newHeight}px`;
    }
  }

  addToggle() {
    /*
    Checks for overflow of box
    Operating on DOM b/c needs to happen after render
    */
    const {
      id,
      type,
    } = this.props;

    const box = window.document.getElementById(`${type}-${id}`);
    let element = box.getElementsByClassName('followup-flow')[0];

    if (type === 'contribution') {
      // eslint-disable-next-line prefer-destructuring
      element = element.getElementsByTagName('div')[1];
    }

    const canOverflow = element.scrollHeight > element.clientHeight;
    if (canOverflow) {
      window.document.getElementById(`${type}-${id}-toggle`).style.display = '';
    } else {
      window.document.getElementById(`${type}-${id}-toggle`).style.display = 'none';
    }
  }

  render() {
    const {
      childAction,
      childCount,
      collapseHandler,
      element,
      id,
      isOpened,
      modalAction,
      parentAction,
      parentCount,
      showLessLabel,
      showMoreLabel,
      type,
    } = this.props;

    return (
      <div
        role="presentation"
        id={`${type}-${id}`}
        className={`followup-timeline-box followup-timeline-box-collapsible${isOpened() ? '' : ' collapsed'}`}
        onClick={modalAction}
      >
        {!!parentCount && (
        <ArrowButton
          direction="inward"
          label={parentCount.toString()}
          onClick={parentAction}
        />
        )}
        {element}
        {!!childCount && (
        <ArrowButton
          direction="outward"
          label={childCount.toString()}
          onClick={childAction}
        />
        )}
        <div
          id={`${type}-${id}-toggle`}
          className="followup-timeline-box-toggle"
        >
          <RaisedButton
            id={`${type}-${id}-button`}
            label={isOpened() ? showLessLabel : showMoreLabel}
            onClick={
              () => {
                collapseHandler();
                // forceUpdate b/c can't pass in state from FollowUpContainer due to resolveElement
                this.forceUpdate();
                if (isOpened()) {
                  window.document.getElementById(`${type}-${id}`).scrollIntoView();
                }
              }
            }
          />
        </div>
      </div>
    );
  }
}

FollowUpBox.defaultProps = {
  modalAction: null,
};

FollowUpBox.propTypes = {
  childAction: PropTypes.func.isRequired,
  childCount: PropTypes.number.isRequired,
  collapseHandler: PropTypes.func.isRequired,
  element: PropTypes.element.isRequired,
  id: PropTypes.number.isRequired,
  isOpened: PropTypes.func.isRequired,
  modalAction: PropTypes.func,
  parentAction: PropTypes.func.isRequired,
  parentCount: PropTypes.number.isRequired,
  showLessLabel: PropTypes.string.isRequired,
  showMoreLabel: PropTypes.string.isRequired,
  type: PropTypes.oneOf(['contribution', 'snippet', 'document']).isRequired,
};

export default FollowUpBox;
