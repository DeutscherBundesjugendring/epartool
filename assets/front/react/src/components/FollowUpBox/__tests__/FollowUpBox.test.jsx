import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import sinon from 'sinon';
import React from 'react';
import injectTapEventPlugin from 'react-tap-event-plugin';
import FollowUpBox from '../FollowUpBox';


injectTapEventPlugin();

describe('rendering', () => {
  it('renders correctly without arrows', () => {
    const tree = shallow(
      <FollowUpBox
        collapseHandler={() => {}}
        id={1}
        type="contribution"
        element={<div>Element</div>}
        isOpened={() => {}}
        parentCount={0}
        parentAction={() => {}}
        childCount={0}
        childAction={() => {}}
        showLessLabel="Show less"
        showMoreLabel="Show more"
      />
    );

    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('renders correctly with right arrow', () => {
    const tree = shallow(
      <FollowUpBox
        collapseHandler={() => {}}
        id={1}
        type="contribution"
        element={<div>Element</div>}
        isOpened={() => {}}
        parentCount={0}
        parentAction={() => {}}
        childCount={1}
        childAction={() => {}}
        showLessLabel="Show less"
        showMoreLabel="Show more"
      />
    );

    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('renders correctly with left arrow', () => {
    const tree = shallow(
      <FollowUpBox
        collapseHandler={() => {}}
        id={1}
        type="contribution"
        element={<div>Element</div>}
        isOpened={() => {}}
        parentCount={1}
        parentAction={() => {}}
        childCount={0}
        childAction={() => {}}
        showLessLabel="Show less"
        showMoreLabel="Show more"
      />
    );

    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('renders correctly with both arrows', () => {
    const tree = shallow(
      <FollowUpBox
        collapseHandler={() => {}}
        id={1}
        type="contribution"
        element={<div>Element</div>}
        isOpened={() => {}}
        parentCount={1}
        parentAction={() => {}}
        childCount={1}
        childAction={() => {}}
        showLessLabel="Show less"
        showMoreLabel="Show more"
      />
    );

    expect(shallowToJson(tree)).toMatchSnapshot();
  });
});

describe('functionality', () => {
  it('calls modal action', () => {
    const spy = sinon.spy();
    const component = shallow(
      <FollowUpBox
        collapseHandler={() => {}}
        id={1}
        type="contribution"
        element={<div>Element</div>}
        isOpened={() => {}}
        parentCount={0}
        parentAction={() => {}}
        childCount={0}
        childAction={() => {}}
        modalAction={spy}
        showLessLabel="Show less"
        showMoreLabel="Show more"
      />
    );

    component.find('div').first().simulate('touchTap', { preventDefault: () => {} });
    expect(spy.calledOnce).toEqual(true);
  });
});
