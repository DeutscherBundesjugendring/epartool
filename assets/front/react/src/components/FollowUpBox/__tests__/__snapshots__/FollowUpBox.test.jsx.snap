// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering renders correctly with both arrows 1`] = `
<div
  className="followup-timeline-box followup-timeline-box-collapsible collapsed"
  id="contribution-1"
  onClick={null}
  role="presentation"
>
  <ArrowButton
    direction="inward"
    disabled={false}
    label="1"
    onClick={[Function]}
  />
  <div>
    Element
  </div>
  <ArrowButton
    direction="outward"
    disabled={false}
    label="1"
    onClick={[Function]}
  />
  <div
    className="followup-timeline-box-toggle"
    id="contribution-1-toggle"
  >
    <RaisedButton
      disabled={false}
      id="contribution-1-button"
      label="Show more"
      onClick={[Function]}
    />
  </div>
</div>
`;

exports[`rendering renders correctly with left arrow 1`] = `
<div
  className="followup-timeline-box followup-timeline-box-collapsible collapsed"
  id="contribution-1"
  onClick={null}
  role="presentation"
>
  <ArrowButton
    direction="inward"
    disabled={false}
    label="1"
    onClick={[Function]}
  />
  <div>
    Element
  </div>
  <div
    className="followup-timeline-box-toggle"
    id="contribution-1-toggle"
  >
    <RaisedButton
      disabled={false}
      id="contribution-1-button"
      label="Show more"
      onClick={[Function]}
    />
  </div>
</div>
`;

exports[`rendering renders correctly with right arrow 1`] = `
<div
  className="followup-timeline-box followup-timeline-box-collapsible collapsed"
  id="contribution-1"
  onClick={null}
  role="presentation"
>
  <div>
    Element
  </div>
  <ArrowButton
    direction="outward"
    disabled={false}
    label="1"
    onClick={[Function]}
  />
  <div
    className="followup-timeline-box-toggle"
    id="contribution-1-toggle"
  >
    <RaisedButton
      disabled={false}
      id="contribution-1-button"
      label="Show more"
      onClick={[Function]}
    />
  </div>
</div>
`;

exports[`rendering renders correctly without arrows 1`] = `
<div
  className="followup-timeline-box followup-timeline-box-collapsible collapsed"
  id="contribution-1"
  onClick={null}
  role="presentation"
>
  <div>
    Element
  </div>
  <div
    className="followup-timeline-box-toggle"
    id="contribution-1-toggle"
  >
    <RaisedButton
      disabled={false}
      id="contribution-1-button"
      label="Show more"
      onClick={[Function]}
    />
  </div>
</div>
`;
